packages/client/src/events/call.ts

```diff
@@ -1,123 +1,44 @@
-import { CallState, StreamVideoWriteableStateStore } from '../store';
+import { CallingState, CallState } from '../store';
 import { StreamVideoEvent } from '../coordinator/connection/types';
+import { Call } from '../Call';
 
 /**
- * Event handler that watched the delivery of CallAcceptedEvent
- * Updates the state store and notifies its subscribers that
- * the given user will be joining the call.
+ * Event handler that watched the delivery of `call.accepted`.
+ * Once the event is received, the call is joined.
  */
-export const watchCallAccepted = (store: StreamVideoWriteableStateStore) => {
-  return function onCallAccepted(event: StreamVideoEvent) {
+export const watchCallAccepted = (call: Call) => {
+  return async function onCallAccepted(event: StreamVideoEvent) {
     if (event.type !== 'call.accepted') return;
-    const { call_cid } = event;
-    const acceptedIncomingCall = store.incomingCalls.find(
-      (incomingCall) => incomingCall.cid === call_cid,
-    );
-    if (acceptedIncomingCall) {
-      console.warn('Received CallAcceptedEvent for an incoming call');
-      return;
+    const { state } = call;
+    if (state.callingState === CallingState.IDLE) {
+      await call.join();
     }
-
-    const acceptedOutgoingCall = store.outgoingCalls.find(
-      (outgoingCall) => outgoingCall.cid === call_cid,
-    );
-    const activeCall = store.activeCall;
-
-    // FIXME OL: we should revisit this logic, it is hard to follow
-    const acceptedActiveCall =
-      activeCall?.cid !== undefined && activeCall.cid === call_cid
-        ? activeCall
-        : undefined;
-
-    if (!acceptedOutgoingCall && !acceptedActiveCall) {
-      console.warn(
-        `CallAcceptedEvent received for a non-existent outgoing call (CID: ${call_cid}`,
-      );
-      return;
-    }
-
-    // once in active call, it is unnecessary to keep track of accepted call events
-    if (call_cid === acceptedActiveCall?.cid) {
-      return;
-    }
-
-    // do not set a new accepted call while in an active call? It would lead to joining a new active call.
-    // todo: solve the situation of 2nd outgoing call being accepted in the UI SDK
-
-    store.setAcceptedCall(event);
   };
 };
 
 /**
- * Event handler that watches delivery of CallRejected Websocket event.
- * Updates the state store and notifies its subscribers that
- * the given user will not be joining the call.
+ * Event handler that watches delivery of `call.rejected` Websocket event.
+ * Once the event is received, the call is left.
  */
-export const watchCallRejected = (store: StreamVideoWriteableStateStore) => {
-  return function onCallRejected(event: StreamVideoEvent) {
+export const watchCallRejected = (call: Call) => {
+  let totalRejections = 0;
```

Comment:
I think on the long run we could handle this better, I've created a Notion task with my suggestions: https://www.notion.so/stream-wiki/8e8d9743b99f4379b24bd2415e5bfdbd?v=714cda30bb2d433fa7151a33556811c5&p=f2994a2225ea48868f69e8167356092f&pm=c - but i think it's fine for now

Reply:
We can definitely do this better... counting the rejections was the simplest solution now.
I like your proposal, we can implement it next/later 👍